In [387]:
#importing the libraries
import pandas as pd
import numpy as np


# Algoritms
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor



# data visualization
import seaborn as sns
sns.set(style="white")
%matplotlib inline
from matplotlib import pyplot as plt
from matplotlib import style

In [389]:
#getting the data
train = pd.read_csv("dataset/testfiles/train.csv")
test = pd.read_csv("dataset/testfiles/test.csv")

In [391]:
train.head()

Unnamed: 0                              NU_INSCRICAO  NU_ANO  \
0           1  ed50e8aaa58e7a806c337585efee9ca41f1eb1ad    2016   
1           2  2c3acac4b33ec2b195d77e7c04a2d75727fad723    2016   
2           3  f4545f8ccb9ff5c8aad7d32951b3f251a26e6568    2016   
3           4  3d6ec248fef899c414e77f82d5c6d2bffbeaf7fe    2016   
4           5  bf896ac8d3ecadd6dba1dfbf50110afcbf5d3268    2016   

   CO_MUNICIPIO_RESIDENCIA NO_MUNICIPIO_RESIDENCIA  CO_UF_RESIDENCIA  \
0                  4314902            Porto Alegre                43   
1                  2304707                  Granja                23   
2                  2304400               Fortaleza                23   
3                  3304557          Rio de Janeiro                33   
4                  1302603                  Manaus                13   

  SG_UF_RESIDENCIA  NU_IDADE TP_SEXO  TP_ESTADO_CIVIL  ...  Q041  Q042  Q043  \
0               RS        24       M              0.0  ...   5.0     A     A   
1               CE        17       F              0.0  ...   NaN     A     A   
2               CE        21       F              0.0  ...   NaN     A     A   
3               RJ        25       F              0.0  ...   5.0     C     A   
4               AM        28       M              0.0  ...   NaN     A     A   

  Q044  Q045 Q046  Q047  Q048  Q049  Q050  
0    A     A    A     A     A     B     D  
1    C     A    B     A     A     C     A  
2    A     A    C     A     A     B     A  
3    A     A    A     D     A     A     A  
4    A     A    A     A     A     A     A  

[5 rows x 167 columns]

In [393]:
test.head()

NU_INSCRICAO  CO_UF_RESIDENCIA  \
0  73ff9fcc02f0a99919906c942c2e1a1042cdcf98                41   
1  71a95f9f1b91a82c65ad94abbdf9f54e6066f968                21   
2  b38a03232f43b11c9d0788abaf060f7366053b6d                23   
3  70b682d9a3636be23f6120fa9d6b164eb3c6002d                15   
4  715494628a50142ce8cb17191cfe6d0f3cae0934                41   

  SG_UF_RESIDENCIA  NU_IDADE TP_SEXO  TP_COR_RACA  TP_NACIONALIDADE  \
0               PR        22       F            3                 1   
1               MA        26       F            3                 1   
2               CE        21       M            1                 1   
3               PA        27       F            3                 1   
4               PR        18       M            1                 1   

   TP_ST_CONCLUSAO  TP_ANO_CONCLUIU  TP_ESCOLA  ...  NU_NOTA_COMP5  \
0                1                5          1  ...           40.0   
1                1                8          1  ...          100.0   
2                2                0          2  ...           80.0   
3                1                8          1  ...            NaN   
4                2                0          2  ...            0.0   

   NU_NOTA_REDACAO  Q001  Q002  Q006  Q024  Q025  Q026  Q027  Q047  
0            420.0     B     A     C     A     A     C     C     A  
1            580.0     E     B     C     B     B     B     F     A  
2            320.0     E     E     D     B     B     A   NaN     A  
3              NaN     H     E     G     B     B     A   NaN     A  
4            320.0     D     H     H     C     B     A   NaN     A  

[5 rows x 47 columns]

In [395]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13730 entries, 0 to 13729
Columns: 167 entries, Unnamed: 0 to Q050
dtypes: float64(28), int64(79), object(60)
memory usage: 17.5+ MB


In [397]:
train.describe()

Unnamed: 0   NU_ANO  CO_MUNICIPIO_RESIDENCIA  CO_UF_RESIDENCIA  \
count  13730.000000  13730.0             1.373000e+04      13730.000000   
mean    2288.173926   2016.0             3.088548e+06         30.736708   
std     1321.230607      0.0             9.890495e+05          9.862405   
min        0.000000   2016.0             1.100023e+06         11.000000   
25%     1144.000000   2016.0             2.307650e+06         23.000000   
50%     2288.000000   2016.0             3.118601e+06         31.000000   
75%     3432.000000   2016.0             3.550308e+06         35.000000   
max     4586.000000   2016.0             5.300108e+06         53.000000   

           NU_IDADE  TP_ESTADO_CIVIL   TP_COR_RACA  TP_NACIONALIDADE  \
count  13730.000000     13202.000000  13730.000000      13730.000000   
mean      21.593809         0.116043      2.185652          1.048871   
std        7.043654         0.365986      1.007285          0.258895   
min       13.000000         0.000000      0.000000          0.000000   
25%       17.000000         0.000000      1.000000          1.000000   
50%       19.000000         0.000000      3.000000          1.000000   
75%       23.000000         0.000000      3.000000          1.000000   
max       67.000000         3.000000      5.000000          4.000000   

       CO_MUNICIPIO_NASCIMENTO  CO_UF_NASCIMENTO  ...         Q032  \
count             1.312100e+04      13121.000000  ...  6354.000000   
mean              3.057946e+06         30.429159  ...     4.530217   
std               9.703020e+05          9.670030  ...     1.027613   
min               1.100023e+06         11.000000  ...     0.000000   
25%               2.306900e+06         23.000000  ...     5.000000   
50%               3.106200e+06         31.000000  ...     5.000000   
75%               3.549805e+06         35.000000  ...     5.000000   
max               5.300108e+06         53.000000  ...     5.000000   

              Q033          Q034          Q035          Q036          Q037  \
count  6354.000000  13730.000000  13730.000000  13730.000000  13730.000000   
mean      3.962858      4.645594      4.678441      3.341515      4.458995   
std       1.625356      0.920998      0.865039      1.711930      1.200172   
min       0.000000      0.000000      0.000000      0.000000      0.000000   
25%       3.000000      5.000000      5.000000      2.000000      5.000000   
50%       5.000000      5.000000      5.000000      4.000000      5.000000   
75%       5.000000      5.000000      5.000000      5.000000      5.000000   
max       5.000000      5.000000      5.000000      5.000000      5.000000   

               Q038          Q039          Q040         Q041  
count  13730.000000  13730.000000  13730.000000  2938.000000  
mean       3.724618      3.506409      4.442098     3.671545  
std        1.689362      1.718858      1.181549     1.746444  
min        0.000000      0.000000      0.000000     0.000000  
25%        3.000000      2.000000      4.000000     3.000000  
50%        5.000000      4.000000      5.000000     5.000000  
75%        5.000000      5.000000      5.000000     5.000000  
max        5.000000      5.000000      5.000000     5.000000  

[8 rows x 107 columns]

In [399]:
pd.DataFrame(abs(train.corr()['NU_NOTA_MT']).sort_values(ascending = False))

NU_NOTA_MT
NU_NOTA_MT                  1.000000
NU_NOTA_CN                  0.584941
NU_NOTA_CH                  0.529594
NU_NOTA_LC                  0.494695
NU_NOTA_REDACAO             0.379376
...                              ...
IN_SALA_ACOMPANHANTE             NaN
IN_MOBILIARIO_ESPECIFICO         NaN
IN_MATERIAL_ESPECIFICO           NaN
TP_PRESENCA_LC                   NaN
TP_PRESENCA_MT                   NaN

[107 rows x 1 columns]

In [401]:
train.columns

Index(['Unnamed: 0', 'NU_INSCRICAO', 'NU_ANO', 'CO_MUNICIPIO_RESIDENCIA',
       'NO_MUNICIPIO_RESIDENCIA', 'CO_UF_RESIDENCIA', 'SG_UF_RESIDENCIA',
       'NU_IDADE', 'TP_SEXO', 'TP_ESTADO_CIVIL',
       ...
       'Q041', 'Q042', 'Q043', 'Q044', 'Q045', 'Q046', 'Q047', 'Q048', 'Q049',
       'Q050'],
      dtype='object', length=167)

In [403]:
test.columns

Index(['NU_INSCRICAO', 'CO_UF_RESIDENCIA', 'SG_UF_RESIDENCIA', 'NU_IDADE',
       'TP_SEXO', 'TP_COR_RACA', 'TP_NACIONALIDADE', 'TP_ST_CONCLUSAO',
       'TP_ANO_CONCLUIU', 'TP_ESCOLA', 'TP_ENSINO', 'IN_TREINEIRO',
       'TP_DEPENDENCIA_ADM_ESC', 'IN_BAIXA_VISAO', 'IN_CEGUEIRA', 'IN_SURDEZ',
       'IN_DISLEXIA', 'IN_DISCALCULIA', 'IN_SABATISTA', 'IN_GESTANTE',
       'IN_IDOSO', 'TP_PRESENCA_CN', 'TP_PRESENCA_CH', 'TP_PRESENCA_LC',
       'CO_PROVA_CN', 'CO_PROVA_CH', 'CO_PROVA_LC', 'CO_PROVA_MT',
       'NU_NOTA_CN', 'NU_NOTA_CH', 'NU_NOTA_LC', 'TP_LINGUA',
       'TP_STATUS_REDACAO', 'NU_NOTA_COMP1', 'NU_NOTA_COMP2', 'NU_NOTA_COMP3',
       'NU_NOTA_COMP4', 'NU_NOTA_COMP5', 'NU_NOTA_REDACAO', 'Q001', 'Q002',
       'Q006', 'Q024', 'Q025', 'Q026', 'Q027', 'Q047'],
      dtype='object')

O dataset train possui 167 columns, enquanto o test possui 47 columns. Como devemos predizer a nota do test vamos utilizar as columas que estão no test e em seguida se necessario dropar as columas necessárias.

In [405]:
#salvando ids para o submission
ids=test['NU_INSCRICAO']

train_colunas_alvo=train[['NU_INSCRICAO', 'CO_UF_RESIDENCIA', 'SG_UF_RESIDENCIA', 'NU_IDADE',
       'TP_SEXO', 'TP_COR_RACA', 'TP_NACIONALIDADE', 'TP_ST_CONCLUSAO',
       'TP_ANO_CONCLUIU', 'TP_ESCOLA', 'TP_ENSINO', 'IN_TREINEIRO',
       'TP_DEPENDENCIA_ADM_ESC', 'IN_BAIXA_VISAO', 'IN_CEGUEIRA', 'IN_SURDEZ',
       'IN_DISLEXIA', 'IN_DISCALCULIA', 'IN_SABATISTA', 'IN_GESTANTE',
       'IN_IDOSO', 'TP_PRESENCA_CN', 'TP_PRESENCA_CH', 'TP_PRESENCA_LC',
       'CO_PROVA_CN', 'CO_PROVA_CH', 'CO_PROVA_LC', 'CO_PROVA_MT',
       'NU_NOTA_CN', 'NU_NOTA_CH', 'NU_NOTA_LC', 'TP_LINGUA',
       'TP_STATUS_REDACAO', 'NU_NOTA_COMP1', 'NU_NOTA_COMP2', 'NU_NOTA_COMP3',
       'NU_NOTA_COMP4', 'NU_NOTA_COMP5', 'NU_NOTA_REDACAO', 'Q001', 'Q002',
       'Q006', 'Q024', 'Q025', 'Q026', 'Q027', 'Q047','NU_NOTA_MT']]

In [407]:
pd.DataFrame(abs(train_colunas_alvo.corr()['NU_NOTA_MT']).sort_values(ascending = False))

NU_NOTA_MT
NU_NOTA_MT                1.000000
NU_NOTA_CN                0.584941
NU_NOTA_CH                0.529594
NU_NOTA_LC                0.494695
NU_NOTA_REDACAO           0.379376
TP_DEPENDENCIA_ADM_ESC    0.373233
TP_LINGUA                 0.201726
TP_COR_RACA               0.157364
CO_UF_RESIDENCIA          0.107493
TP_ST_CONCLUSAO           0.087869
TP_ENSINO                 0.086353
TP_ESCOLA                 0.069304
NU_IDADE                  0.061485
TP_STATUS_REDACAO         0.059376
TP_NACIONALIDADE          0.034965
TP_ANO_CONCLUIU           0.029039
IN_TREINEIRO              0.022287

In [409]:
#dropar colunas inuteis
train_colunas_alvo.drop(['NU_INSCRICAO','SG_UF_RESIDENCIA','IN_DISCALCULIA','TP_PRESENCA_CN','TP_PRESENCA_CH','IN_DISLEXIA','IN_GESTANTE','IN_BAIXA_VISAO','IN_CEGUEIRA','TP_PRESENCA_LC','IN_SABATISTA','IN_SURDEZ','IN_IDOSO','NU_NOTA_COMP3','NU_NOTA_COMP5','NU_NOTA_COMP4','NU_NOTA_COMP2','NU_NOTA_COMP1','CO_PROVA_CH','CO_PROVA_CN','CO_PROVA_MT','CO_PROVA_LC'], axis=1, inplace=True)
test.drop(['NU_INSCRICAO','SG_UF_RESIDENCIA','IN_DISCALCULIA','TP_PRESENCA_CN','TP_PRESENCA_CH','IN_DISLEXIA','IN_GESTANTE','IN_BAIXA_VISAO','IN_CEGUEIRA','TP_PRESENCA_LC','IN_SABATISTA','IN_SURDEZ','IN_IDOSO','NU_NOTA_COMP3','NU_NOTA_COMP5','NU_NOTA_COMP4','NU_NOTA_COMP2','NU_NOTA_COMP1','CO_PROVA_CH','CO_PROVA_CN','CO_PROVA_MT','CO_PROVA_LC'], axis=1, inplace=True)

KeyError: "['NU_INSCRICAO' 'SG_UF_RESIDENCIA' 'IN_DISCALCULIA' 'TP_PRESENCA_CN'\n 'TP_PRESENCA_CH' 'IN_DISLEXIA' 'IN_GESTANTE' 'IN_BAIXA_VISAO'\n 'IN_CEGUEIRA' 'TP_PRESENCA_LC' 'IN_SABATISTA' 'IN_SURDEZ' 'IN_IDOSO'\n 'NU_NOTA_COMP3' 'NU_NOTA_COMP5' 'NU_NOTA_COMP4' 'NU_NOTA_COMP2'\n 'NU_NOTA_COMP1' 'CO_PROVA_CH' 'CO_PROVA_CN' 'CO_PROVA_MT' 'CO_PROVA_LC'] not found in axis"

In [412]:
pd.DataFrame(abs(train_colunas_alvo.corr()['NU_NOTA_MT']).sort_values(ascending = False))

NU_NOTA_MT
NU_NOTA_MT                1.000000
NU_NOTA_CN                0.584941
NU_NOTA_CH                0.529594
NU_NOTA_LC                0.494695
NU_NOTA_REDACAO           0.379376
TP_DEPENDENCIA_ADM_ESC    0.373233
TP_LINGUA                 0.201726
TP_COR_RACA               0.157364
CO_UF_RESIDENCIA          0.107493
TP_ST_CONCLUSAO           0.087869
TP_ENSINO                 0.086353
TP_ESCOLA                 0.069304
NU_IDADE                  0.061485
TP_STATUS_REDACAO         0.059376
TP_NACIONALIDADE          0.034965
TP_ANO_CONCLUIU           0.029039
IN_TREINEIRO              0.022287

In [414]:
train_colunas_alvo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13730 entries, 0 to 13729
Data columns (total 26 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   CO_UF_RESIDENCIA        13730 non-null  int64  
 1   NU_IDADE                13730 non-null  int64  
 2   TP_SEXO                 13730 non-null  object 
 3   TP_COR_RACA             13730 non-null  int64  
 4   TP_NACIONALIDADE        13730 non-null  int64  
 5   TP_ST_CONCLUSAO         13730 non-null  int64  
 6   TP_ANO_CONCLUIU         13730 non-null  int64  
 7   TP_ESCOLA               13730 non-null  int64  
 8   TP_ENSINO               4282 non-null   float64
 9   IN_TREINEIRO            13730 non-null  int64  
 10  TP_DEPENDENCIA_ADM_ESC  4282 non-null   float64
 11  NU_NOTA_CN              10341 non-null  float64
 12  NU_NOTA_CH              10341 non-null  float64
 13  NU_NOTA_LC              10133 non-null  float64
 14  TP_LINGUA               13730 non-null

In [416]:
train_simples=train[['TP_SEXO','NU_NOTA_CN','NU_NOTA_CH','NU_NOTA_LC','NU_NOTA_REDACAO',	'TP_DEPENDENCIA_ADM_ESC','TP_COR_RACA','CO_UF_RESIDENCIA','NU_NOTA_MT']]
test_simples=test[['TP_SEXO','NU_NOTA_CN','NU_NOTA_CH','NU_NOTA_LC','NU_NOTA_REDACAO',	'TP_DEPENDENCIA_ADM_ESC','TP_COR_RACA','CO_UF_RESIDENCIA']]

In [418]:
train_simples.fillna(0,inplace=True)
test_simples.fillna(0,inplace=True)

In [420]:
total = train_simples.isnull().sum().sort_values(ascending=False)
percent_1 = train_simples.isnull().sum()/train_simples.isnull().count()*100
percent_2 = (round(percent_1, 1)).sort_values(ascending=False)
missing_data = pd.concat([total, percent_2], axis=1, keys=['Total', '%'])
missing_data.head()

Total    %
NU_NOTA_MT                  0  0.0
CO_UF_RESIDENCIA            0  0.0
TP_COR_RACA                 0  0.0
TP_DEPENDENCIA_ADM_ESC      0  0.0
NU_NOTA_REDACAO             0  0.0

In [422]:
train_cat = train_simples['TP_SEXO']
test_cat = test_simples['TP_SEXO']
train_cat_encoded = pd.get_dummies(train_cat)
test_cat_encoded = pd.get_dummies(test_cat)

KeyError: 'TP_SEXO'

In [425]:
train_simples.drop('TP_SEXO', axis=1, inplace=True)
test_simples.drop('TP_SEXO', axis=1, inplace=True)
train_num = train_simples.join(train_cat_encoded)
test_num = test_simples.join(test_cat_encoded)

KeyError: "['TP_SEXO'] not found in axis"

In [428]:
train_num.describe()

NU_NOTA_CN    NU_NOTA_CH    NU_NOTA_LC  NU_NOTA_REDACAO  \
count  13730.000000  13730.000000  13730.000000     13730.000000   
mean     356.621515    398.923911    381.166737       390.447633   
std      213.282315    237.174594    234.648409       267.749316   
min        0.000000      0.000000      0.000000         0.000000   
25%      365.625000    345.900000      0.000000         0.000000   
50%      433.100000    499.600000    486.000000       480.000000   
75%      493.200000    564.400000    548.300000       580.000000   
max      806.400000    807.000000    763.600000      1000.000000   

       TP_DEPENDENCIA_ADM_ESC   TP_COR_RACA  CO_UF_RESIDENCIA    NU_NOTA_MT  \
count            13730.000000  13730.000000      13730.000000  13730.000000   
mean                 0.704734      2.185652         30.736708    356.092607   
std                  1.117178      1.007285          9.862405    228.844193   
min                  0.000000      0.000000         11.000000      0.000000   
25%                  0.000000      1.000000         23.000000      0.000000   
50%                  0.000000      3.000000         31.000000    422.500000   
75%                  2.000000      3.000000         35.000000    504.400000   
max                  4.000000      5.000000         53.000000    952.000000   

                  F             M  
count  13730.000000  13730.000000  
mean       0.618427      0.381573  
std        0.485790      0.485790  
min        0.000000      0.000000  
25%        0.000000      0.000000  
50%        1.000000      0.000000  
75%        1.000000      1.000000  
max        1.000000      1.000000

In [429]:
from sklearn.model_selection import train_test_split
target = train_simples['NU_NOTA_MT']
predictors = train_num.drop(['NU_NOTA_MT'], axis=1)

x_train, x_val, y_train, y_val = train_test_split(predictors, target, test_size = 0.2, random_state = 0)

from xgboost import XGBRegressor

my_model = XGBRegressor()
# Add silent=True to avoid printing out updates with each cycle
my_model.fit(x_train, y_train, verbose=False)

[01:23:55] WARNING: src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0,
             importance_type='gain', learning_rate=0.1, max_delta_step=0,
             max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
             n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
             reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
             silent=None, subsample=1, verbosity=1)

In [430]:
my_model.score(x_train, y_train)
acc_my_model = round(my_model.score(x_train, y_train) * 100, 2)
print(acc_my_model)

93.03


In [431]:
# make predictions
predictions = my_model.predict(test_num)

In [432]:
submission = pd.DataFrame({ 'NU_INSCRICAO' : ids, 'NU_NOTA_MT': predictions.astype('int64') })
submission.to_csv('answer.csv', index=False)